### ISSUE TITLE
"Power in" sends electricity back, basically working as a relay for junction boxes

### ISSUE THREAD

**Issue by hobotam** [Regalis11/Barotrauma#9603]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

I wanted to make the battery to distribute electricity to the limited amount of sources, just essentials to reduce power consumption. I used relay to divide the wires and connected all to "power in" slots. To that slots also connected wires from main junction boxes, so the problem is, for some reason on turning off the reactor and battery on, it powers the whole sub, even things that seemingly are not connected to it. By brute force testing I determined that "power in" for some reason works as just "power", like junction boxes and lamps, testing further I found out that in previous versions of the game everything was ok and "power in" worked normally.
Strangely, I haven't found reports on this issue, so I'm writing my own.
![image](https://user-images.githubusercontent.com/64069664/179633362-58c4aa46-ec12-49cb-bae9-16b5fd26ed66.png)

### Reproduction steps

1. Create a test chamber in the sub editor (I assume you got plenty of them already)
2. Add any power output, battery or reactor
3. Connect power output to any object that has "power in" in it
4. Add junction box and connect it's "power" to "power in" of that object
5. Add another object that consumes power and connect it to junction box
6. Go into test, turn on the power output, junction box while not connected to the battery/reactor now receives power from that "power in" wire, and spreading it further, meaning that "power in" now works like relay but only for junction boxes

### Bug prevalence

Happens every time I play

### Version

0.18.15.0

### Which operating system did you encounter this bug on?

Windows

**Comment by NilanthAnimosus** [comment]
I am not sure I understand from the description alone, but in my experience the pin on "Power_In" for relays acts like a junction box normally and same for "Power_Out" on relays but I have not checked this behavior for many versions now nor with none relay objects which are very different to relays.

Your picture shows only one on power in, yet you say everything connected to power in? meanwhile you also say in the reproduction steps that you mean the power in of that object, so I assume like a pump with two wires going to it and not anything to strictly do with the relays themselves but the power_in pin of a power consuming object acts like a junction box in your tests?

**Comment by hobotam** [comment]
> 

Sorry I shouldn't have written this long story of how I encountered the bug, it's kinda useless in the report and turned out to be confusing.
What I meant was, by powering any object that has "power in" wire slot from any source, it also sends electricity to connected junction box, that can further power other objects, which doesn't seem right and how "power in" should work.
I recorded a short video of me testing it, in words, battery is powering oxygen generator that is also connected to junction box, by removing the wire you can clearly see that the power goes through oxygen generator.

https://user-images.githubusercontent.com/64069664/179663051-7e7f3506-a76b-4233-86db-fc35b53c00ab.mp4

What I didn't do is check how much power goes through the object to junction box but from the looks of it there's no limit.

**Comment by NilanthAnimosus** [comment]
Looking into this I am no longer sure if this is a bug or intended behavior, as a power rework occurred that changed the general behavior of the power network (see: https://github.com/Regalis11/Barotrauma/pull/8184) and has likely been present since version 0.17.0.0.

it looks intended for relays at least to do this. not sure on daisy-chaining devices from the power input pin however.

**Comment by hobotam** [comment]
> it looks intended for relays at least to do this. not sure on daisy-chaining devices from the power input pin however.

I wouldn't reported it if I didn't find it inconvenient >:(
I don't want for my pumps to power my junction boxes and by that to power my whole electrical system, making it impossible, or at least very complicated, to power just a portion of a sub (what I essentially wanted to do).

**Comment by NilanthAnimosus** [comment]
I understand the use case, and to be honest don't personally think its intended or even a good idea. It basically makes junction boxes worthless and relay's fairly difficult to use intentionally, just adding unnecessary complication and yeah - inconvenience for any backup-based system intending to wire a device to two separate sources of power.  I would be surprised though if this wasn't addressed since February as it does look like it was known in the pull request (Possibly was patched then reintroduced at some point?)

**Comment by Regalis11** [comment]
This was actually an intentional change done as a part of the power rework: wires connected to the same pin are now considered part of the same grid. This doesn't apply to wires between two power_in connections or two power_out connections, so it isn't possible to just daisy-chain all your devices together and bypass the need for junction boxes or relays. This change vastly simplified the power logic and got rid of a ton of bugs and inconsistencies in the electrical system.

Consider a case such as yours where two different grids are supplying power to a device. What would the voltage of each of the grids be? How do the grids "share" the load of the device between each other? I'd argue these kinds of questions made the old system much more unintuitive (and buggy) than the current one.

>I don't want for my pumps to power my junction boxes and by that to power my whole electrical system, making it impossible, or at least very complicated, to power just a portion of a sub (what I essentially wanted to do).

Now if you want to split the grids into two, you need to use a relay. For example in that latest video you posted, you need to put a relay between the junction box and the oxygen generator to prevent power from propagating from the other grid to the junction box.

**Comment by NilanthAnimosus** [comment]
Apparently the issue here from the looks of it is a device (IE. a pump) is connecting two grids separated by a relay.

Does your description @Regalis11 mean you need two relays to get battery power from one, reactor power from another, and not to cross the power from either when both connect to the same receiver (IE. oxygen gen/pump) ?

the implication here is two separate power networks (from a relay) are being connected by a pump's power in

**Comment by Regalis11** [comment]
Yes, that's correct. Wires connected in the same pin are part of the same grid, except if the wire is run between the power_in connections or two power_out connections, which isn't the case here.

**Comment by hobotam** [comment]
> Now if you want to split the grids into two, you need to use a relay. For example in that latest video you posted, you need to put a relay between the junction box and the oxygen generator to prevent power from propagating from the other grid to the junction box.

I do hope that it did fix things and maybe made it simpler for others even though power in/out are still called in/out while not actually being just in/out. But oh well, now I need to create a relay for each individual object that I want to be "separated" from reactor and battery grids, in my case it's 5 new relays, which is not a lot, but man do I not envy those that have a lot bigger subs with 2 or even 3 separated grids. All of those relays would have to be hidden in game or just in walls too cause let's be honest it would look horrible having a relay component near each of the pumps. 
Sorry for the rant, relay is a painful topic for me :'(

**Comment by Regalis11** [comment]
> I do hope that it did fix things and maybe made it simpler for others even though power in/out are still called in/out while not actually being just in/out.

I'd say that depends on how you look at it. The way I see it, the power_in pin of a device still only feeds power in to the device, but the two power wires going into it are obviously connected to each other, so there's no reason why power wouldn't be able to move between them.

> but man do I not envy those that have a lot bigger subs with 2 or even 3 separated grids

That depends on how they've been built: if they're actually separate grids (as in, not connected to each other), it shouldn't require any changes. I would also question the need to isolate the grids from each other completely: in your specific case, why not have one grid that can be fed power from the reactor, batteries or both?

**Comment by hobotam** [comment]
> That depends on how they've been built: if they're actually separate grids (as in, not connected to each other), it shouldn't require any changes. I would also question the need to isolate the grids from each other completely: in your specific case, why not have one grid that can be fed power from the reactor, batteries or both?

In my case my sub is small and I got place for only one battery that can't really power the whole thing and not be out in seconds. That's why I wanted it to be an emergency power source that would only power essentials like pumps, engine and control panel so in case that I'm at full speed forward and my reactor runs out of fuel I can save myself with it. It also works as a standby mode, so when I go outside my sub I can put it on battery, while standing still it consumes almost no power which saves a lot of fuel and makes sure that my sub will be there when I come back.

**Comment by KrakenTales** [comment]
As I understand it, after the discussion was closed by the developer, the discussion thread did not continue productively. I will add my opinion here, if it somehow affects further development.

> The way I see it, the power_in pin of a device still only feeds power in to the device, but the two power wires going into it are obviously connected to each other, so there's no reason why power wouldn't be able to move between them.

I can agree with this only in cases where it really works, namely, when I described here #9603. 
Due to the lack of logic according to the physics of processes or human expectation, I fervently hope that this feature will be corrected, or let all intersecting wires transmit power further, no matter where they intersect, whether it be power_in pin or power_out pin, as required by physics and expected in general.
